cupy/cuda/stream.pyx

```diff
@@ -178,32 +208,35 @@ class BaseStream(object):
         # This operator is implemented to compare the singleton instance
         # of null stream (Stream.null) can safely be compared with null
         # stream instance created by a user.
-        return self.ptr == other.ptr
+        if self.device_id == -1 or other.device_id == -1:
+            return self.ptr == other.ptr
+        else:
+            return (self.device_id == other.device_id
+                    and self.ptr == other.ptr)
```

Comment:
This is actually a bug fix that I think we should backport...For example, multiple devices can have stream 88 created, but they should not compare to equal.

Reply:
Okay, I'll make a PR that backports this part after #5172 is merged.

Reply:
Sorry, I couldn't get the point here. I think it's safe to assume that different streams have different pointers?

Reply:
It's the case for `Stream.null` and `Stream.ptds`.

Reply:
Right, because they are "global" singletons, not "per-device" singletons.

Reply:
Wait. As @kmaehashi pointed out, can `cudaStreamCreate` create the same stream pointer for different devices?

Reply:
As for singletons, device_id is always -1 and ptr is always 1 or 2, so I think device_id comparison is not necessary.

Reply:
Ohhh yeah, I got it now, sorry you're right @kmaehashi, I will revert this part.